subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java

```diff
@@ -66,7 +66,12 @@ public DefaultJavaPluginExtension(JavaPluginConvention convention,
         this.project = project;
         this.modularity = objectFactory.newInstance(DefaultModularitySpec.class);
         this.jvmPluginServices = jvmPluginServices;
-        this.toolchain = objectFactory.newInstance(DefaultToolchainSpec.class);
+        DefaultToolchainSpec toolchainSpec = objectFactory.newInstance(DefaultToolchainSpec.class);
+        this.toolchain = toolchainSpec;
+        if (convention instanceof DefaultJavaPluginConvention) {
```

Comment:
I'd actually do a blind cast. If we ever have another convention implementation, it must be handled differently and not ignored (as it would be right now).